### PR TITLE
New formalism for `_Def_Fields` and `_model` file

### PR DESCRIPTION
**Pull request by DaluS**
## First the miscellaneous modification 

1. Changed again (a merge might have not functionned properly) the "Hub" name of the core
2. Modified a part of the _PRINT_OR_WAIT_ to function correctly in the case of float number ( some others are still not working) 
3. Added a basic function to describe each model and the presets in __init__ of models called `_core._class_checks.models.describe_available_models()`
4. 
image

## Modification on `_def_fields_NEWFORMALISM` 

Correspond to modification of #59 , the file itself is not added in the workflow but is ready to. 
The main properties : 
* `_DLIBRARY` is the one accessible for the modeler (easy to modify/implement new fields
* The group classification is caused by the organization of the library ( dict of dict ), allow folding in IDE also. 
* Dimensions are now a dictionnary with in keys : 
    * multiplier (to handle Gigaton, 10^12 etc after)
    * each unit (y, dollars etc) are keys and the exponent is the value ( example productivity )
image
* Ode, state function, parameters have a different writing 
    * parameters 
image

    * ode  
image

    * functions 
image

* There will be functions added to transform a parameter into a state function (easy), a parameter into an ODE (easy), a state function into an ODE (more complex)

## MODIFICATIONS ON `_model_GK-NEWFORMALISM`

* _DPARAM will be transformed into _LOGICS, as a dictionnary of dictionnaries, with one section for `ode`, one for state functions `func`, other fields for DDE and PDE will be added later when needed
* `_FUNC_ORDER` and `_FUNC_PARAMS` will not need to be filled

## WHAT TO COME  

* Transforming CORE so that he is able to read both files
* Adding the overloading elements to the class 
* Adding overload-parrallel system (transformation to N systems in parrallel) 
* Enjoying it as it will be fully functional with a consistent formalism that allows future development :)

### CODE REVIEW

**Comment by Didou09** [comment]
> Changed again (a merge might have not functionned properly) the "Hub" name of the core

As mentionned in the PR #56  itself: you did not merge, you closed the PR, without merging, that's why

File: models/_model_GK-NEWFORMALISM.py [hunk 1]
```diff
@@ -0,0 +1,150 @@
...
+
+import numpy as np
+
+
+# ---------------------------
+
+
+_FUNC_ORDER = None # user-defined function order (optional)
+
+_FUNC_PARAMS = None # user-defined list of parameters (optional)
```
Comment (Didou09) [inline models/_model_GK-NEWFORMALISM.py:16]: But you don't specify the list of parameters here for that model, where is it specified?

The user might not want to specify all parameters in the preset, only those the values of which he wants to change

There should be somewhere a table listing the parameters for each model, if it is not here, where do you want to put it ?
We could try to decipher it dynamically from the functions if you don't want to have it stored, that's a possibility I guess

File: models/_def_fields_NEWFORMALISM.py [hunk 2]
```diff
@@ -0,0 +1,389 @@
+# -*- coding: utf-8 -*-
```
Comment (Didou09) [inline models/_def_fields_NEWFORMALISM.py:1]: What's the difference between this file (_def_fields_NEWFORMALISM.py) and the previous one (_def_fields2.py) ? Which one is the new one ?

**Review by Didou09 — COMMENTED** [review]
I guess this is work in progress since the tests are not passing.
So I am not approving the PR at this point, just providing an intermediate review with comments / questions.

I see you point in the model files, you want to separate the set of equations from the parameters and initial values. I agree this is a good move.

However, I don't see (yet) where you want to define the list of parameters for a given model?

Also, I would probably remove remove the field 'value' from _def_fields.py, in order to separate what is likely to be changed by the user ('value') from is likely not to be changed ('units', com'), so the user does not try to edit _def_fields.py but only the model file (the presets).

This would allow to define the defaults values in model (where the user can change them) at the same time as defining the list of parameters.

What do you think ? Maybe a visio today afternoon would help the discussion ?

**Comment by DaluS** [comment]
> > Changed again (a merge might have not functionned properly) the "Hub" name of the core
> 
> As mentionned in the PR #56 itself: you did not merge, you closed the PR, without merging, that's why

Oh I didn't understood that ! 
It's same issue for the unit test not working : they try to test the class solver and not hub (I'll fix that) 

> However, I don't see (yet) where you want to define the list of parameters for a given model?
> But you don't specify the list of parameters here for that model, where is it specified?
>  There should be somewhere a table listing the parameters for each model, if it is not here, where do you want to put it ?
We could try to decipher it dynamically from the functions if you don't want to have it stored, that's a possibility I guess

It should not be necessary to precise parameters. When checking the model (all the lambda functions and all the keys in the dictionnary), we can determine the functions (called in lambda, in keys) and parameters (called in lambda, not in keys). 
That might be dangerous I agree, so maybe using the previous parameters definition (then check consistency with above elements) might be safer (but more tedious for modeler). a `_PARAMETERS`.

image

> The user might not want to specify all parameters in the preset, only those the values of which he wants to change

As a preset overload `_dparam`, we only have to state the points we want to modify ! That's exactly the idea and the formalism for the moment in 

> What's the difference between this file (_def_fields_NEWFORMALISM.py) and the previous one (_def_fields2.py) ? Which one is the new one ?

I will remove `_def_fields2` it's depreciated :) I forgot about it ! 


> Also, I would probably remove remove the field 'value' from _def_fields.py, in order to separate what is likely to be changed by the user ('value') from is likely not to be changed ('units', com'), so the user does not try to edit _def_fields.py but only the model file (the presets).
> This would allow to define the defaults values in model (where the user can change them) at the same time as defining the list of parameters.

I think the library should have default value ( putting ode is not a good idea for example) so that a user that has no Idea of one value could use this default one.

**Comment by Didou09** [comment]
@Dalus, ok for a meeting tomorrow,

Ho by the way, I don't want to emphasize only the points that deserve discussion, this new feature (`_core._class_checks.models.describe_available_models()`) you introduce is a very good idea I think, congratulations ! (sorry, I should have mentionned it earlier)

**Comment by Didou09** [comment]
> Oh I didn't understood that !

Remember you can always re-open the [PR](https://github.com/DaluS/GEMMES/pull/56) and merge it, but first you need to restore the branch (because apparently you deleted it).

If you want we can do this together tomorrow

**Comment by DaluS** [comment]
This is suprising, it looks like I broke `_model_GK`, (the rest is working), I don't know what happened :) 

I'll check that tomorrow ! No need to merge (it's a draft)

--------------------------------

What's in : 

* An update to the readme
* A few updates on the `_NEWFORMALISM` that is ready to be inserted into the system
* An ensemble of functions I'd like to implement in the core 

What's next : 
* Repair `_model_GK`
* finish the PR 
* rewrite the necessary part of the `_core` to implement the newformalism
* add the functionalities that are in `cockpit` in the comment section 

```
for NameOfTheModel in ListOfModels :
    sol = _core.Hub(NameOfTheModel)
    PRESETS = []
    for preset in PRESETS :
        sol = _core.Hub(NameOfTheModel)
        
        #sol.load_preset(preset)
        
        sol.load_field({'a' : 1})
        sol.load_field({'a' : [1,2,'lin']})
        sol.load_field({'a' : [1,2,'log'],
                        'alpha' : 0.03})
        sol.load_field({'nx': 100})   
        sol.load_field({'a': [1,2,3]}) 
        
        sol.get_summary()
        sol.run(solver='eRK4-homemade', verb=1.1)
        
        sol.getCycleAnalysis(key='lambda')
        sol.getCycleAnalysis(key=False)
        
        Result = sol.get_dparam(returnas=dict)
                
        plt.figure()
        plt.subplot(211);plt.plot(Result['time']['value'],Result['omega']['value']);plt.ylabel('Omega')
        plt.subplot(212);plt.plot(Result['time']['value'],Result['lambda']['value']);plt.ylabel('Lambda')
        plt.suptitle(NameOfTheModel+' '+preset)
        plt.show()
        
        sol.plot('all')
        
```

**Comment by Didou09** [comment]
Pro tip: to solve conflicts (because devel has changed a bit), go on your local devel branch, update it (git pull origin devel), then go on your local branch NewFormalism and merge devel into it (git merge devel), so you'll solve conflicts locally, then push to the remote NewFormalism branch

**Comment by DaluS** [comment]
The reason why it is not working is because the previous initial condition was very bad - W=0.1 implied that 13% of the initial production went to salary in that case, which caused investment to be huge and send the system in a weird place. 
In consequence, the system is saying that the result differ which is true : the initial conditions in the program (with a depreciated writing in the future) are different, but better to test the system :)

**Pull request closed without merging**